index.d.ts

```diff
@@ -0,0 +1,193 @@
+import * as http from 'http';
+
+export = Router;
+
+declare namespace Router {
+
+  interface RouterOptions {
+    strict?: boolean;
+    caseSensitive?: boolean;
+    mergeParams?: boolean;
+  }
+
+  interface Layer {
+    name: string,
+    handle: RequestHandler | ErrorRequestHandler,
+    handle_request: RequestHandler,
+    handle_error: ErrorRequestHandler,
+    match: (path: string) => boolean,
+  }
+
+  interface IncomingRequest extends http.IncomingMessage {
+    originalUrl?: string,
+    params?: {
+      [key: string]: any,
+    },
+  }
+
+  interface RoutedRequest extends IncomingRequest {
+    baseUrl: string,
+    next?: NextFunction,
+    route?: IRoute
+  }
+
+  type RequestParamHandler = (req: IncomingRequest, res: http.ServerResponse, next: NextFunction, value: string, name: string) => void;
```

Comment:
I tried to figure out where it went wrong in the TS, and I think this line is the culprit: inside of the `router.param` callback, there is `req.params`, `req.route`, and all the other things available. I think that this should be the "routed request" type here.

Reply:
Changed that to `RoutedRequest`.